syntax = "proto3";

/*
 * Buffer streams exchanged between client and server are comprised of the
 * following sequence of encoded (serialized) protocol buffer messages (except
 * for the first byte, which represents the length of the encoded header
 * message).
 *
 * +-----------+----------------------------+--------------------+============================================+
 * | header len| Encoded RqHeader/RsHeader  | Encoded CryptoInfo || Encoded, padded and encrypted Rq/Rs body ||
 * | (1 byte)  | (variable length < 256)    | (variable length)  || (variable length)                        ||
 * +-----------+----------------------------+--------------------+============================================+
 *
 * The Rq/Rs body is padded (AES has a 16-byte block length) and encrypted
 * after serialization and decrypted and de-padded before deserialization. The
 * header length byte, header and CryptoInfo fields are not encrypted. 
*/

message RqHeader {
    int32 partner_id  = 1;
    int32 crypto_info_length = 2;
    int32 rq_length = 3; // Includes any AES padding
}

message RsHeader {
    int32 crypto_info_length = 1;
    int32 rs_length = 2; // Includes any AES padding
}

message CryptoInfo {
    bytes iv=1; // Always 16 bytes in length
    int32 aes_padding_length = 2;
}

// Message body definitions.
//
message Rq {
    bytes device_id = 1;
    Aps aps = 2;
    GsmCells gsm_cells = 3;
    UmtsCells umts_cells = 4;
    LteCells lte_cells = 5;
    CdmaCells cdma_cells = 6;
    NbiotCells nbiot_cells = 7;
}

message Aps {
    uint32 connected_ap_idx_plus_1 = 1; // 0 means that device is not connected to any AP
    uint32 common_age_plus_1 = 2; // Valued only if all age values are the same.
    repeated int64 mac = 3;
    repeated int32 channel_number = 4;
    repeated uint32 neg_rssi = 5; // -rssi
    repeated uint32 age = 6; // Not valued if all ages are the same. common_age_plus_1 sent instead in this case.
}

message GsmCells {
    repeated uint32 mcc = 1;
    repeated uint32 mnc = 2;
    repeated uint32 lac = 3;
    repeated uint32 ci = 4;
    repeated sint32 neg_rssi = 5; // -rssi
    repeated uint32 age = 6;
}

message UmtsCells {
    repeated uint32 mcc = 1;
    repeated uint32 mnc = 2;
    repeated uint32 lac = 3;
    repeated uint32 ci = 4;
    repeated sint32 neg_rscp = 5; // -rscp
    repeated uint32 age = 6;
}

message CdmaCells {
    repeated uint32 sid = 1;
    repeated uint32 nid = 2;
    repeated uint32 bsid = 3;
    repeated sint32 neg_rssi = 4; // -rssi
    repeated uint32 age = 5;
}

message LteCells {
    repeated uint32 mcc = 1;
    repeated uint32 mnc = 2;
    repeated uint32 tac = 3;
    repeated uint32 eucid = 4;
    repeated sint32 neg_rsrp = 5; // -rsrp
    repeated uint32 age = 6;
}

message NbiotCells {
    repeated uint32 mcc = 1;
    repeated uint32 mnc = 2;
    repeated uint32 tac = 3;
    repeated uint32 e_cellid = 4;
    repeated uint32 neg_nrsrp = 5; // -nrsrp
    repeated uint32 age = 7;
}

message Rs {
    double lat = 1;
    double lon = 2;
    float hpe = 3;
}
